hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/CloudObjectsSelectorCommon.java

```diff
@@ -0,0 +1,118 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.utilities.sources.helpers;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hudi.common.config.SerializableConfiguration;
+import org.apache.hudi.common.fs.FSUtils;
+import org.apache.hudi.exception.HoodieException;
+import org.apache.hudi.exception.HoodieIOException;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.spark.api.java.function.FlatMapFunction;
+import org.apache.spark.sql.Row;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Generic helper methods to fetch from Cloud Storage during incremental fetch from cloud storage buckets.
+ * NOTE: DO NOT use any implementation specific classes here. This class is supposed to across S3EventsSource,
+ * GcsEventsSource etc...so you can't assume the classes for your specific implementation will be available here.
+ */
+public class CloudObjectsSelectorCommon {
+
+  private static final Logger LOG = LogManager.getLogger(CloudObjectsSelectorCommon.class);
+
+  /**
+   * Return a function that extracts filepaths from a list of Rows.
+   * Here Row is assumed to have the schema [bucket_name, filepath_relative_to_bucket]
+   * @param storageUrlSchemePrefix Eg: s3:// or gs://. The storage-provider-specific prefix to use within the URL.
+   * @param serializableConfiguration
+   * @param checkIfExists check if each file exists, before adding it to the returned list
+   * @return
+   */
+  public static FlatMapFunction<Iterator<Row>, String> getCloudFilesPerPartition(
+          String storageUrlSchemePrefix, SerializableConfiguration serializableConfiguration, boolean checkIfExists) {
+    return rows -> {
+      List<String> cloudFilesPerPartition = new ArrayList<>();
+      rows.forEachRemaining(row -> {
+        addFileToList(row, cloudFilesPerPartition, storageUrlSchemePrefix, serializableConfiguration, checkIfExists);
+      });
+
+      return cloudFilesPerPartition.iterator();
+    };
+  }
+
+  /**
+   * Extract a filepath from a given Row, and add it to a list. Optionally check if the file exists.
+   * Here Row is assumed to have the schema [bucket_name, filepath_relative_to_bucket]
+   * @param storageUrlSchemePrefix Eg: s3:// or gs://. The storage-provider-specific prefix to use within the URL.
+   */
+  private static void addFileToList(Row row, List<String> filePathUrls, String storageUrlSchemePrefix,
```

Comment:
Can you return just the output file instead of passing filePathUrls?

Reply:
Didn't understand the question. What do mean by return the output file - a Java File object?
The filePathUrls will be of the form gs://bucket-name/folder-name/file.json. At most this function checks if the file exists, for which it needs the gcs-connector type libs on the classpath. There's a separate part of the code which fetches the data of the file.

I'm updating the Javadoc of that function to be like below. Do you think this is sufficient?
  /**
   * Construct a fully qualified file path from a given Row, and add it to a list. Optionally check if the file exists.
   * The checkIfExists part assumes the relevant impl classes for the storageUrlSchemePrefix are already present
   * on the classpath!
   * Here Row is assumed to have the schema [bucket_name, filepath_relative_to_bucket]
   * @param storageUrlSchemePrefix Eg: s3:// or gs://. The storage-provider-specific prefix to use within the URL.
   */

Reply:
Sorry, I meant to change the function signature to :
  String addFileToList(Row row, String storageUrlSchemePrefix, ...)
ie remove filePathUrls from input params and rather return the filePath. 
And this applies to addCloudFile() as well. Also then the function name should be changed appropriately.

Reply:
I think I'll hold off on this for now. If time permits I might do another pass and pick this up. Let's see.

Reply:
Done. Latest commit returns a filePathUrl instead of taking in input list and modifying it.